### PR TITLE
clients: agent surfaces can edit clients, in the product's own vocabulary

### PR DESCRIPTION
**Pull request by chelojimenez**
The inspector half of "robust Clients APIs on the agent surfaces" — workstreams A (naming), the client-facing half of B (robust editing), and C (exposure). The backend half is [MCPJam/mcpjam-backend#1149](https://github.com/MCPJam/mcpjam-backend/pull/1149) and **must be deployed first**: the canonical `/clients` routes call `hosts:resolveHostByNameOrId` and `hosts:updateHostFields`. `@mcpjam/sdk` must likewise only publish after this is deployed, since its new methods call `/clients`.

## The problem

The UI's noun is **Client** — "a named, reusable configuration that defines how MCPJam connects to and talks to your MCP servers". Every agent surface spoke backend vocabulary instead (`/hosts`, `list_hosts…delete_host`, `mcpjam cloud hosts …`), and the two surfaces agents actually live on — the MCP catalog and the agent operation registry — excluded client editing entirely by written policy. Editing was also not robust: the only edit was a whole-config replace with no partial patch and no concurrency token.

## Naming (workstream A)

`/api/v1/projects/:projectId/clients` is canonical, with `Client*` schemas and a `Clients` tag; the operations are `list_clients`…`duplicate_client`; the CLI is `mcpjam cloud clients …` with `.alias("hosts")` and `--client` accepting `--host` (both is a usage error, not a precedence rule — precedent: `--repetitions`/`--iterations`).

The `/hosts` routes, `cloud hosts` commands and `listHosts…duplicateHost` SDK methods stay as deprecated compatibility surfaces **on their original shapes**. That is why the compatibility types are distinct interfaces rather than aliases: `/hosts` returns `hostConfigId` where `/clients` returns `configId`, so a type alias would be a compile-time lie about a runtime shape and would break typecheck for every caller reading a field the route still sends. Aliases carry `Deprecation: true`, stay ID-only, and keep the pre-rename tokenless write contract — a compatibility surface that starts rejecting requests it used to accept is a break wearing a deprecation notice.

The deprecated `*_host` operations stay executable but are **not** in `ALL_OPERATIONS`. Every registered surface partitions that list, so leaving them out is what guarantees no old name can be advertised as a tool or persisted in a proposal row.

Nothing in the AI-host-compat concept is renamed (`/host-catalog`, `check_host_compatibility`, `hostStyle`), nor `clientCapabilities`/`clientInfo` inside a config — those are the MCP-protocol client, a different thing — nor anything inside the opaque config blob. Renaming the `host` selector on compose/eval/environment ops is deferred: those are live external wire fields.

## Robust editing (workstream B, client side)

`PATCH /clients/:client` gains `set`: named fields applied over the client's current config inside the write transaction, where absent means keep and `null` means reset (required field) or clear (optional one).

Every config-affecting write requires `expectedConfigId` and every rename requires `expectedName` — two tokens, because a rename does not rotate the config and config identity is blind to a concurrent one. The config token is content-addressed, so the same id means byte-identical settings. A stale token is a `409` whose `details` carries the current value, so recovery needs no second round-trip.

The canonical detail path takes a client **name or id**, resolved server-side so one implementation owns the ambiguity and eligibility rules; a narrowly-scoped list-and-scan fallback covers only a genuinely absent backend function, never an authorization refusal or an ambiguity error. Private User Testing backing clients are excluded by default from listing, from name resolution **and** from id lookup, because the rule is about what a client *is* and cannot lapse for ids.

## Exposure (workstream C)

| Op | MCP catalog | Agent registry | Chat toolset |
|---|---|---|---|
| `list_clients` / `get_client` | added | direct | excluded |
| `create_client` / `update_client` / `set_client_servers` | added | **gated** | excluded |
| `duplicate_client` | added | excluded | excluded |
| `delete_client` | excluded | excluded | excluded |

This reverses the written "infrastructure writes" exclusions deliberately. The line that replaces them is **bounded, preconditioned overwrite versus resource removal**: an overwrite names what it replaces, is refused outright if the client changed since the caller read it, and leaves the client standing. Deletion does none of that.

Annotations stay honest rather than convenient: create and duplicate are additive (`risk: "none"` → `destructiveHint: false`); update and set-servers replace settings currently in force (`risk: "destructive"` → `destructiveHint: true`) and are catalog-visible anyway, behind compare-and-set. The comment equating `destructiveHint` with permanent deletion is updated — the taxonomy always said "removes or invalidates something that existed"; deletion just happened to be the whole membership until now. The version trail is deliberately **not** cited as an undo path: there is no public restore operation today.

The three gated entries are named `TIER_EXCEPTIONS` in the derivation test, in both directions (create deviates from `risk:none → direct`, the other two from `risk:destructive → excluded`), because that test exists to force exactly this review.

`freezeClientWriteArgs` pins three things at proposal-mint time, fail-closed via `requiredFrozenKeys`:
1. **the target** — name → exact id, so a rename cannot repoint an approved edit;
2. **the tokens** — verified against a server read, never substituted: freezing in whatever the server currently has would turn compare-and-set into "overwrite whatever is there";
3. **the impact** its approval copy quotes — a real operation field, checked transactionally, so a consumer attached between proposing and clicking makes the write conflict instead of widening what a human agreed to.

The idempotency projection drops the display-only label (a harmless rename must not mint a duplicate approval) and keeps the impact (a changed blast radius *should* mint a new one). Copy branches on the actual edit: a rename-only proposal says exactly "Rename client **Claude** to **Claude QA**" and claims nothing about execution, while a config edit enumerates all three counts — including when they are all zero, since "this affects nothing else" is information and omitting the sentence reads as the counts having been left out.

The in-app chat toolset keeps all seven excluded: the Clients tab and the WebMCP client tools own that surface, and a chat tool editing the client the chat runs on would be a worse version of the editor already on screen.

## Verification

- **Full inspector suite**: 17,928 passed / 51 skipped across 1,429 files.
- **CLI**: 1,073 passed, 0 failed (includes a new `clients-set-parser.test.ts` covering every field family).
- **SDK**: 6,323 passed / 8 skipped.
- **MCP**: 53 passed (catalog order, README table, and all four annotation cases).
- Typecheck clean for `@mcpjam/sdk`, `@mcpjam/cli`, `@mcpjam/mcp`; `check:platform-runtime-safety` clean.

New/rewritten tests worth calling out:

- `clients.test.ts` (61) — canonical DTOs incl. `configId`/`impact`, the token requirements and their 400s, a backend `CONFLICT` surfacing as a 409 with `currentConfigId`, private-backing filtering by name **and** by id, name resolution plus the narrow skew fallback (and proof an authorization failure is *not* treated as skew), and a dedicated alias suite pinning `hostConfigId`, the old three-field detail shape, the absent read-backs, tokenless writes, and ID-only resolution.
- `client-write-freeze.test.ts` (17) — name→id freeze, token verification refusing to mint, hidden-backing and missing-impact refusals, the `requiredFrozenKeys` declaration, idempotency identity under a rename vs. under a changed impact, and the four copy branches.
- `--set` parser tests pin the cases a generic JSON-or-string guess gets wrong: `systemPrompt=123` stays a string, `temperature=0.2x` fails locally, `requireToolApproval=1` is refused rather than silently becoming `false`, and `--unset modelId` fails without leaving the machine.

Docs: OpenAPI rewritten for the seven operations (aliases baselined in `KNOWN_UNDOCUMENTED` with a reason, which is what that list is for), plus `docs/cli/reference.mdx` — including the pre-existing gap where `servers` and `duplicate` were never documented — `docs/cli/overview.mdx`, `docs/cli/migration.mdx` and `docs/reference/public-api.mdx`.

## Not done here

Workstream E5's staging end-to-end run needs both repos deployed together, so it is not something this PR can carry out; the CLI and MCP scenarios it calls for are written up in the plan. The optional WebMCP `ui_*_client` renames are left as an independent change, per the plan's decision.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01GtVBMirwx7fTfi3mq4xYdn

---
_Generated by [Claude Code](https://claude.ai/code/session_01GtVBMirwx7fTfi3mq4xYdn)_

---

> [!NOTE]
> **Medium Risk**
> Broad changes to REST, CLI, SDK, and MCP around client configuration and concurrent edits; blast radius includes live environments and journeys, though compare-and-set and gated proposals reduce silent overwrite risk and deprecated paths preserve prior behavior.
> 
> **Overview**
> Replaces **host** vocabulary on agent-facing surfaces with **client** — the product term for named MCP connection configurations — while keeping **`/hosts`**, **`mcpjam cloud hosts`**, and legacy SDK host methods as deprecated aliases on their **original** DTOs and tokenless write shapes.
> 
> **Canonical API and CLI.** New **`/api/v1/projects/:projectId/clients`** routes (OpenAPI `Clients` tag) expose list/get/create/update/delete, server replacement, and duplicate. **`mcpjam cloud clients`** mirrors that surface with **`hosts`** as a command alias and **`--client`** / deprecated **`--host`** (both together is a usage error). **`clients update`** adds field-owned **`--set`/`--unset`** parsing and requires **`expectedConfigId`** for any config change and **`expectedName`** for renames; **`clients servers`** requires **`expectedConfigId`**. Stale tokens surface as **`409`** with current values in **`details`**.
> 
> **Robust editing on the server.** Canonical **`PATCH`** supports a transactional **`set`** block (absent = keep, **`null`** = reset/clear) alongside whole-config replace; writes forward compare-and-set tokens to Convex. Detail paths resolve **name or ID**, hide User Testing backing clients by default, and return **`configId`**, **`impact`**, and related read-backs on the canonical shape only.
> 
> **Agent exposure.** MCP catalog adds **`list_clients`**, **`get_client`**, and bounded writes (**`create_client`**, **`update_client`**, **`set_client_servers`**, **`duplicate_client`**); **`delete_client`** stays excluded. Agent registry gates create/update/set-servers with **`freezeClientWriteArgs`** (pinned id, verified tokens, frozen **`impact`** for approval copy). **`update_client`** / **`set_client_servers`** advertise **`destructiveHint: true`** as deterministic overwrites, not deletions.
> 
> Docs, op-bindings, SDK/OpenAPI parity maps, and extensive tests cover alias compatibility, **`--set`** parser edge cases, private-backing filtering, and gated proposal copy/idempotency.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit f96efe5f6578860324e799f1fdf1152115e782a8. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by chelojimenez** [comment]
### :white_check_mark: **Snyk checks have passed. No issues have been found so far.**

| Status | Scan Engine |  Critical |  High |  Medium |  Low | Total (0) |
|---|---|---|---|---|---|---|
| :white_check_mark: | Open Source Security| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/marcelo-apk6cgUZZztPZRjPdWkYwH/pr-checks/2908aed0-e034-4b3d-a433-1bd021987301?refs=f96efe5f6578860324e799f1fdf1152115e782a8&source=prChecksComment) |

:computer: Catch issues earlier using the plugins for [VS Code](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner), [JetBrains IDEs](https://plugins.jetbrains.com/plugin/10972-snyk-security), [Visual Studio](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner-vs-2022), and [Eclipse](https://marketplace.eclipse.org/content/snyk-security#external-install-button).

File: mcpjam-inspector/server/routes/v1/clients.ts [hunk 1]
```diff
@@ -0,0 +1,1024 @@
...
+
+  // Trim on BOTH branches: the normalization belongs to the write boundary, not
+  // to one of the two ways of reaching it. A template is authored data too, and
+  // one carrying a padded `modelId` would otherwise persist an id that no
+  // downstream verbatim comparison recognizes.
+  const input = withTrimmedModelId(
+    body.template
+      ? await resolveHostTemplateInput(body.template, body.theme)
+      : body.config!
+  );
```
[the remark this answers is not in the document]
Comment (chelojimenez) [reply mcpjam-inspector/server/routes/v1/clients.ts:754]: Declining this one — I checked `main` before deciding, and neither mint path got stricter.

`main`'s `hosts.ts` already enforced the forward-client invariant on both:

- `createHostSchema`'s `superRefine` carried the identical `hostConfigPinsAModel(value.config!)` check (`main`, `hosts.ts:257`), and
- the duplicate handler already refused with the same message: *"Host ... does not pin a model, so duplicating it would create another host that cannot back a headless environment"* (`main`, `hosts.ts:445-449`).

So the shared schema and the shared refusal preserve the alias's behavior exactly — there is no request that used to succeed on `/hosts` and now fails. Gating them on `!surface.legacy` would be the actual regression in the other direction: it would open a hole where a legacy caller could mint a modelless client that `create` refuses, and those clients fail much later, at launch, with `ENV_MODEL_REQUIRED`.

The two places that genuinely needed lenient schemas are the ones that gained *new* requirements — `updateHostSchema` and `updateHostServersSchema`, which stay tokenless — and those did get them.

Everything else in this review is fixed in e1102a7.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: docs/reference/openapi.json [hunk 2]
```diff
@@ -22,8 +22,8 @@
   ],
   "tags": [
     {
-      "name": "Hosts",
-      "description": "Project hosts: named model + capability profiles you run chats and eval suites against."
+      "name": "Clients",
+      "description": "Clients — the named, reusable configurations that define how MCPJam connects to and talks to your MCP servers. The original `/hosts` paths remain as deprecated, ID-only compatibility aliases with their original DTOs and their original (tokenless) write contracts; every alias response carries `Deprecation: true`. New integrations should use `/clients`."
```
[the remark this answers is not in the document]
Comment (chelojimenez) [reply docs/reference/openapi.json]: Declining — the second half of your ask ("or state that these aliases are intentionally undocumented") is already done, in the two places this repo puts that decision.

1. `__tests__/openapi-drift.test.ts` baselines all seven alias routes in `KNOWN_UNDOCUMENTED`, each with the reason written out. That list exists precisely for "endpoints that exist but are not part of the public contract", and its own comment says an entry without a reason is a backlog item wearing a decision's clothes. The test enforces both directions, so if these were ever documented the entries would have to go.

2. The `Clients` tag description says it in the spec itself: *"The original `/hosts` paths remain as deprecated, ID-only compatibility aliases with their original DTOs and their original (tokenless) write contracts; every alias response carries `Deprecation: true`. New integrations should use `/clients`."*

Publishing both spellings is the thing I'd want to avoid: the spec is what a new integration reads, and describing a deprecated surface there presents a choice where there isn't one. Existing callers don't need the spec to keep working, and they get `Deprecation: true` on every response, which reaches proxies and client libraries that a body field wouldn't.

Your other openapi findings were right and are fixed in e1102a7 — the `ClientServersRequest.expectedConfigId` description that had inherited `ClientUpdateRequest`'s `config`/`set` conditioning, the two "host"s left in `ClientCreateRequest.config.modelId`, and the duplicate request body, which is now `required: false`.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by chelojimenez** [comment]
## CI status on `0d8bb06`: everything green except one CodeQL alert I can't read

**Green:** Inspector Tests 1/4–4/4, E2E Smoke (Playwright, including the OAuth Debugger e2e), Run Tests (SDK → CLI verification), Build and Test, both Snyk checks, Analyze (actions), Analyze (javascript-typescript). Backend counterpart [MCPJam/mcpjam-backend#1149](https://github.com/MCPJam/mcpjam-backend/pull/1149) is green on test and lint.

**Red:** the `CodeQL` rollup check — *"1 new alert including 1 high severity security vulnerability."*

### What I fixed

Two alerts surfaced across the three pushes, and one of them posted an annotation I could read: **Replacement of a substring with itself** ([alert 105](https://github.com/MCPJam/inspector/security/code-scanning/105)) at `cli/tests/clients-set-parser.test.ts`. That was mine and it was real — `key.replace("__", "__")`, a half-written attempt to escape a key before interpolating it into a `RegExp`. Removed in `0d8bb06` by dropping the `RegExp` construction entirely, and that alert is gone from the current head.

Separately, while auditing my own diff for high-severity patterns I found and fixed a genuine prototype-chain bug in the same file's field table (`--set constructor=x` passed the "known field?" guard and returned `undefined`, crashing the caller) — own-property lookup plus an `Object.create(null)` target, in `9f172f4`.

### What I can't establish, and why

The remaining **1 high** is unchanged across all three pushes — `606e991`, `9f172f4` and `0d8bb06` each report exactly one, and neither of the fixes above moved it. Two further facts about it:

- **It posted no annotation as a review comment.** GitHub attaches those to lines inside the PR diff, so this alert's location is not a line this PR changed. Everything I added *is* in the diff (the renamed files are recorded as adds), so it isn't in my new code.
- **CodeQL says so itself**, in the same summary: *"Alerts not introduced by this pull request might have been detected because the code changes were too large."*

Taken together, the best-supported reading is a pre-existing alert on `main` surfaced by the size of this diff — most likely following one of the two renamed files, whose contents moved verbatim. I checked that specifically: `loadConfigObject` in `cli/src/commands/clients.ts`, including both `readFileSync` calls, is **byte-identical** to what shipped in `cli/src/commands/hosts.ts` (verified with a diff of the extracted function; empty output). Git recorded the rename as delete-plus-add, so CodeQL sees those lines as new.

I want to be clear that this is inference, not a reading of the alert. **What blocks confirming it:** this session has no code-scanning-alerts API, the check run's `output.text` is empty, the `Analyze` job logs report only "Exported results to SARIF" without listing alerts, the workflow run publishes no artifact, and CodeQL here runs via GitHub's default setup so there's no workflow file to inspect or re-run. I also can't re-run the check, and wouldn't expect a re-run to change a deterministic static analysis.

### What I'd need

Someone with Security-tab access can settle it in a few seconds at [the PR's alert list](https://github.com/MCPJam/inspector/security/code-scanning?query=pr%3A4378+tool%3ACodeQL+is%3Aopen). If it turns out to be in code this PR actually introduced, point me at it and I'll fix it. If it's the moved `readFileSync` — a CLI reading a path its own operator typed on the command line — then it's the pre-existing alert following the rename, and the honest options are to dismiss it as it presumably already is on `main`, or to address it separately rather than widen this PR with a suppression for code it only moved.

I'm still watching this PR and will keep checking back until it's green or merged.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by chelojimenez** [comment]
## On the latest review's three "outside diff range" findings

Thanks — the formatting one was right, and chasing it turned up something worse than trailing commas.

### 1. Prettier (`client.ts` / `operations.ts`) — real, and much bigger than reported

Confirmed: `npx prettier --check` fails on both. But the fix is not to remove the commas it flags. Digging in:

- `sdk/.prettierrc` sets `trailingComma: "es5"`, while the SDK is actually written in the `"all"` style, and **nothing in CI runs prettier over these paths**. So the files have never conformed to their own config, on `main` or anywhere else.
- Which means my running `prettier --write` on the files I was editing produced **~1,300 lines of pure comma churn in code this PR has nothing to do with** — 444 lines in `client.ts`, 691 in `operations.ts`, plus the agent registry, the MCP catalog and the convex error translator. That was mine and it was noise.

Undone in `2484527` by restoring each file from `main` and re-applying only the semantic edits — not by reformatting again with a different flag, which would have had the same problem pointing the other way. Deletions across the PR went **2,824 → 1,511**.

Chasing the same class of mistake into `openapi.json` found another 329 lines: re-dumping the spec with `ensure_ascii=False` had rewritten all 301 `—` escapes as literal em-dashes. Fixed in `10f324a`; removals in that file went **440 → 111**, and all 111 are now inside the host→client replacement.

Total diff is now **4,664 / 1,182**, down from 6,352 / 2,824. Every deletion left is one this PR meant to make.

Making `sdk/.prettierrc` and the SDK's actual style agree is a real cleanup, but it is a whole-file rewrite of the SDK and belongs in its own PR, not smuggled into this one.

### 2 & 3. The decision-summary findings are about `main`, not this PR

`readEvalRunDecisionSummary`'s cursor fallback and the unconstrained `decision` schema both come from [#4374](https://github.com/MCPJam/inspector/pull/4374) ("Evals D9: one canonical run decision summary"), commit `b26dfa7`, which is on `main` and reached this branch through a merge. Verified:

```
$ git log --oneline -1 -S "readEvalRunDecisionSummary" -- sdk/src/platform/operations.ts
b26dfa7 Evals D9: one canonical run decision summary across API, MCP and CLI (#4374)

$ git diff origin/main -- sdk/src/platform/operations.ts | grep -c "readEvalRunDecisionSummary\|diagnosticsCursor"
0
```

Zero hits — neither line is in this PR's diff against `main`. The review picked them up because it walked my merge commit, which replays `main`'s changes.

Both look like fair points on their own terms, particularly the pagination one: a canonical 404 with a non-empty cursor returning no `decisionSummary` is a real gap. But they're changes to someone else's eval feature, and fixing them here would widen a clients PR into eval-verdict territory and put the fix somewhere its author wouldn't look. Worth raising against `main` or with #4374's author.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by chelojimenez** [comment]
## CodeQL is green — and the churn cleanup is why

Correcting my earlier comment: [I asked for someone with Security-tab access to identify the CodeQL high](https://github.com/MCPJam/inspector/pull/4378#issuecomment-5419982673). **That is no longer needed.** On `10f324a` the check reports *"No new alerts in code changed by this pull request."*

It resolved itself when I removed the reformatting churn, and that is the explanation rather than a coincidence. CodeQL's PR check reports new alerts **in code changed by the pull request**. My `prettier --write` sweep had rewritten ~1,300 lines of code this change never touched, which made those lines "changed by this PR" and pulled a pre-existing alert living in them into scope. Take the churn back out, and the alert falls back out of scope with it.

That is exactly what CodeQL's own hedge was saying the whole time — *"Alerts not introduced by this pull request might have been detected because the code changes were too large"* — and it matches the two things I could establish about the alert: it never moved across three pushes despite unrelated fixes, and it never posted an annotation, because its location was not a line I had meaningfully changed.

Worth being clear about what this is and isn't. Nothing was suppressed and no alert was dismissed. Whatever it is still exists on `main`, unchanged, and if it is real it deserves its own fix — it simply isn't this PR's, and this PR is no longer claiming to have touched it.

### Full status on `10f324a`

| | |
|---|---|
| CodeQL | ✅ no new alerts |
| Inspector Tests 1/4–4/4 | ✅ |
| Run Tests (SDK → CLI verification) | ✅ |
| E2E Smoke (Playwright + OAuth Debugger) | ✅ |
| Build and Test (typecheck packages + client, build, production start) | ✅ |
| Analyze (actions / javascript-typescript) | ✅ |
| Snyk (both orgs) | ✅ |
| Mintlify | ✅ |

Locally, after the restore-and-reapply surgery: full inspector suite **17,966 passed / 51 skipped**, SDK 6,416, CLI 1,088, MCP 53, SDK build and typecheck clean.

Backend counterpart [MCPJam/mcpjam-backend#1149](https://github.com/MCPJam/mcpjam-backend/pull/1149) is green on test and lint, with 9,822 backend tests passing. **It must deploy before this merges** — the canonical `/clients` routes call `hosts:resolveHostByNameOrId` and `hosts:updateHostFields`, and `@mcpjam/sdk` should publish only after this is live, since its new methods call `/clients`.

Both remaining review threads are resolved, including the two I'd pushed back on — thanks for re-checking the create/duplicate one against `main` and withdrawing it.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by chelojimenez** [comment]
### On the `types.ts` union formatting (last review, outside-diff)

Checked, and declining — with numbers, since the merge-risk banner still lists "formatting" as unresolved and I don't want that read as something this PR introduced.

Both flagged lines are **byte-identical to `main`** and are not in this PR's diff. Swapping `main`'s copy of the file in reproduces the two errors exactly:

```
$ npx eslint src/platform/types.ts          # this branch
  434:40  error  Replace `·"direct"·|·"scenario"·|·"eval"` …  prettier/prettier
  864:39  error  Replace `·"emulated"·|·"mixed"` …            prettier/prettier
✖ 2 problems

# same file from origin/main → identical two errors
```

They are 2 of **4,586 pre-existing eslint errors in `sdk/src`**, and **no CI job runs the SDK's lint** — which is why the whole package has drifted from its own config. Fixing two of them here would clear nothing measurable and would put unrelated formatting back into a diff I just spent two commits taking it out of.

This is the same finding as [the earlier prettier one](https://github.com/MCPJam/inspector/pull/4378#issuecomment-5420117099), narrowed to two lines: these are exactly the lines my `prettier --write` had "fixed" and I deliberately reverted to match `main`. Reverting them was the point.

Making `sdk/.prettierrc` and the SDK's actual style agree — and putting `npm run lint -w @mcpjam/sdk` in CI so it stays agreed — is worth doing, and it is a ~4,500-error PR of its own.

CI on `10f324a` is green across the board, CodeQL included.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by chelojimenez** [comment]
`Inspector Tests 1/4` went red on `f96efe5`, on a test this PR does not touch. Re-run queued; here is the diagnosis so it is on the record either way.

**What failed**

```
FAIL server routes/v1/__tests__/tools-call.test.ts
  > POST /v1/projects/:projectId/servers/:serverId/tools/call
  > returns the MCP CallToolResult plus additive durationMs
AssertionError: expected 4 to be greater than or equal to 5
  ❯ routes/v1/__tests__/tools-call.test.ts:78:29
```

**Why it is not this PR's**

The only commit on this head beyond the previously-all-green `10f324a` is a merge of `origin/main` that brings in exactly one file — `.changeset/releases-2026-08-25.md`, a changesets release note. No code, no config, no test. `git diff 10f324a HEAD` is 7 added lines of markdown. Neither `routes/v1/tools.ts` nor `tools-call.test.ts` appears anywhere in this PR's diff, and the test file's last change was #4305.

**Root cause — a latent flake in the assertion, not a regression**

The test has `executeTool` await `setTimeout(…, 5)` and then asserts the route's measured duration is `>= 5`. The route measures with `Math.max(0, Date.now() - startedAt)` (`routes/v1/tools.ts:52`). That assertion assumes `setTimeout(n)` guarantees `Date.now()` advances by at least `n`, which libuv does not promise: timers are scheduled against the loop's cached millisecond clock and can fire fractionally early, and `Date.now()` truncates to whole milliseconds, so the two truncations can compound in the same direction.

Probed directly on this runner — 20,000 iterations of exactly what the test does:

```
samples: 20000 | observed < 5ms: 119 (0.60%)
distribution: [ [4, 119], [5, 15825], [6, 3972], [7, 20], [8, 13], [9, 9] ]
```

So ~0.6% of executions of this one assertion fail with no code change at all, and a loaded CI runner makes that worse. Running the test file eight times in a row locally on this branch reproduced it: **1 of 8 runs failed with the identical `expected 4 to be greater than or equal to 5`**. The other eight assertions in the file are deterministic; only this one is wall-clock-dependent.

**Proposed patch — not applied here, to keep this diff scoped**

The intent of the test is that `durationMs` is measured and additive, which is worth keeping; it just should not be pinned to real elapsed time. The file already stubs the clock in the very next test, so the same technique makes this one deterministic and actually tests the arithmetic:

```ts
const executeTool = vi.fn().mockResolvedValue({
  content: [{ type: "text", text: "ok" }],
});
stubConnection(executeTool);
const now = vi.spyOn(Date, "now");
now.mockReturnValueOnce(1_000).mockReturnValueOnce(1_007);

try {
  const res = await postToolsCall();
  expect(res.status).toBe(200);
  const body = (await res.json()) as { content?: unknown[]; durationMs?: number };
  expect(body.content).toEqual([{ type: "text", text: "ok" }]);
  expect(body.durationMs).toBe(7);
  expect(executeTool).toHaveBeenCalledWith("s1", "echo", {});
} finally {
  now.mockRestore();
}
```

That is a one-test change in a file unrelated to this PR, so I have left it out rather than widening the diff — happy to push it here if a reviewer would rather it ride along, or to open it as its own PR.

---
_Generated by [Claude Code](https://claude.ai/code)_
